cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputFormat.java

```diff
@@ -101,19 +100,22 @@ public void write(K key, V value) throws IOException, InterruptedException {
           String finalPath = relativePath + "/" + outputName;
 
           // if we don't have the record writer yet for the final path, create one and add it to the cache
-          rw = getBaseRecordWriter(getTaskAttemptContext(job, finalPath));
+          TaskAttemptContext taskAttemptContext = getTaskAttemptContext(job, finalPath);
+          rw = getBaseRecordWriter(taskAttemptContext);
           this.recordWriters.put(partitionKey, rw);
+          this.contexts.put(partitionKey, taskAttemptContext);
         }
         rw.write(key, value);
       }
 
       @Override
       public void close(TaskAttemptContext context) throws IOException, InterruptedException {
         try {
-          List<RecordWriter<?, ?>> recordWriters = new ArrayList<>();
-          recordWriters.addAll(this.recordWriters.values());
-          MultipleOutputs.closeRecordWriters(recordWriters, context);
+          Map<PartitionKey, RecordWriter<?, ?>> recordWriters = new HashMap<>();
+          recordWriters.putAll(this.recordWriters);
+          MultipleOutputs.closeRecordWriters(recordWriters, contexts);
 
+          this.contexts.clear();
```

Comment:
I think its ok to not clear contexts.
But if we are clearing contexts, then we should also clear recordWriters...